File: src/main/java/chapter03/test_Recursion.scala
package chapter03

object test_Recursion {
  def main(args: Array[String]): Unit = {
    println(fact(5))
  }
  def fact(n: Int): Int = {
    if(n == 0) return 1
    fact(n - 1) * n
  }
  //尾递归实现
  def tailfact(n: Int): Int = {
    def loop(n: Int, curRes: Int): Int = {
      if(n == 0) return curRes
      loop(n - 1, curRes)
    }
    loop(n, 1)
  }
}

File: src/main/java/test01/test_Recursion.java
package test01;

public class test_Recursion {
    public static void main(String[] args) {
        System.out.println(fact(5));
        System.out.println(fact1(5));
    }
    public static int fact(int n){
        int re = 1;
        for(int i = 1; i <= n; i++){
            re *= i;
        }
        return re;
    }
    public static int fact1(int m){
        if(m == 0) return 1;
        return m * fact1(m - 1);
    }
}
